pkg/controller/backupbucket/controller.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backupbucket
+
+import (
+	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// FinalizerName is the backupbucket controller finalizer.
+	FinalizerName = "extensions.gardener.cloud/backupbucket"
+	// ControllerName is the name of the controller
+	ControllerName = "backupbucket-controller"
+)
+
+// AddArgs are arguments for adding a BackupBucket controller to a manager.
+type AddArgs struct {
+	// Actuator is a BackupBucket actuator.
+	Actuator Actuator
+	// ControllerOptions are the controller options used for creating a controller.
+	// The options.Reconciler is always overridden with a reconciler created from the
+	// given actuator.
+	ControllerOptions controller.Options
+	// Predicates are the predicates to use.
+	// If unset, GenerationChangedPredicate will be used.
+	Predicates []predicate.Predicate
+}
+
+// DefaultPredicates returns the default predicates for a controlplane reconciler.
+func DefaultPredicates(mgr manager.Manager) []predicate.Predicate {
+	return []predicate.Predicate{
+		extensionscontroller.GenerationChangedPredicate(),
+		//TODO: Do we need force deletion? extensionscontroller.AnnotationsChangedPredicate(),
```

Comment:
Hm, good question. Although we might not often use it it would be a nice feature to have, right?!

Reply:
Cool. I'll implement that.

Reply:
It's no more needed. Since deletionGracePeriod logic is completely moved to GCM. The extension/backupEntry resource is marked for deletion by gcm-core/backupEntry controller only after deletionGracePeriod is over. So, the annotation support is on gcm-core/backupEntry rather thant extension/backupEntry Its deletionGracePeriod ism better kept generic rather than extension specific.